src/MDOP.jl

```diff
@@ -87,3 +220,134 @@ function beta_statistic(Y::Dataset, s::Vector; τs = 0:50 , w::Int = 1)
     β = mean(log10.(Δϕ); dims=1)
     return vec(β)
 end
+
+
+"""
+    embed2(Y::Dataset,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+embedding vector `Y`.
+"""
+function embed2(Y::Dataset,s::Vector{Float64},τ::Int)
+    N = size(Y,1)   # length of input trajectory
+    NN = size(Y,2)  # dimensionality of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,NN+1)
+    # fill vector up until index M
+    @inbounds for i = 1:NN
+        Y_new[:,i] = Y[1:M,i]
+    end
+    # add lagged component of s
+    Y_new[:,NN+1] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
+
+"""
+    embed2(y::Vector,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+time series `y`.
+"""
+function embed2(y::Vector{Float64},s::Vector{Float64},τ::Int)
+    N = size(y,1)   # length of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,2)
+    # fill vector up until index M
+    Y_new[:,1] = y[1:M]
+    # add lagged component of s
+    Y_new[:,2] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
+
+
+"""
+    fnn(NNdist,NNdistnew,r::Float64=2.0) -> FNNs
+fnn computes the amount of false nearest neighbors when adding another component
+to a given (vector-) time series. This new component is the `τ`-lagged version
+of a univariate time series. 'NNdist' is storing the distances of the nearest
+neighbor for all considered fiducial points and 'NNdistnew' is storing the
+distances of the nearest neighbor for each fiducial point in one embedding
+dimension higher using a given `τ`. The obligatory threshold `r` is by default
+set to 2.
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+"""
+function fnn(NNdist::T,NNdistnew::T;r::Float64=2.0) where {T}
+    @assert length(NNdist) == length(NNdistnew) "Both input vectors need to store the same number of distances."
+
+    # convert array of arrays into simple vectors, since we only look at K=1
+    NN_old = zeros(length(NNdist))
+    NN_new = zeros(length(NNdistnew))
+    for i = 1:length(NNdist)
+        NN_old[i]=NNdist[i][1]
+        NN_new[i]=NNdistnew[i][1]
+    end
+
+    # first ratio
+    ratio1 = 1/r    # since we input only z-standardized data with unit
+    # construct statistic
+    ratio2 = NN_new./NN_old
+
+    cond1 = ratio2 .> r
+    cond2 = NN_old .< ratio1
+
+    fnns = sum(cond1.*cond2)
+    fnns2 = sum(NN_old .< ratio1)
+
+    # store fraction of valid nearest neighbors
+    if fnns2 == 0
+        FNN = NaN;
+    else
+        FNN = fnns/fnns2
+    end
+
+    return FNN
+end
```

Comment:
In the file `estimate_dimension.jl` we have a function called `fnn`. 

1. The two functions cannot have the same name because they overwrite each other.
2. Can you please give some comments on your method versus the existing one? If yous is better we should use it instead in general. 
3. Can you comment on the **performance** (computing time) of the two functions?

Reply:
> 1. The two functions cannot have the same name because they overwrite each other.

Okay, I'll fix this. Is there any handy command, which tells you whether there is already a function in that package, which uses the same name?

> 2. Can you please give some comments on your method versus the existing one? If yous is better we should use it instead in general.

The existing methods in `estimate_dimension.jl` perform a false nearest neighbor for a given time series, using uniform time delay embedding and a predefined range of encountered dimensions. Here we need something else. We need just the basic idea of the FNN-method, namely counting the number of nearest neighbour distances, which exceed a certain threshold, when passing from one dimension into the next one. 
> when passing from one dimension into the next one.

in this context means to perform an embedding cycle with a delay, which is not known a priori. Consequently, the input to "my fnn"-function, are `Arrays` of nearest neighbour distances for dimension _**d**_ and for dimension _**d+1**_. That's it. Note that there is no nearest neighbour finding computations whatshowever. 

> 3. Can you comment on the performance (computing time) of the two functions?

As seen from my comments above, it does not make sense to compare a "full method" to my "in place, core" method.
If you want, we could draft "my fnn inplace function" to `estimate_dimension.jl`, if you want. It is handy, anyway. Furthermore it uses the idea of [Hegger & Kantz (1999)](https://doi.org/10.1103/PhysRevE.60.4970), whereas your functions `fnn` and `f1nn` use the original method of [Kennel et al. (1992)](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.45.3403) (not so recommended in my opinion) and the method proposed by [Krakovska et al. (2015)](https://doi.org/10.1155/2015/932750), which I also not like that much.

Reply:
> Is there any handy command, which tells you whether there is already a function in that package, which uses the same name?

Yes, `isdefined(Module, :function_name)`.

---

For the remaining comments: okay, I see now. Sure, if you believe you have the time to implement a "full" version, feel free to do it. I also do not like the Krakovska method (in fact it is outperformed by methods published decades earlier).
For now you can focus on finishing this PR though (taking baby steps), and simply renaming `fnn` to something else, more fitting.